tmk_core/common/chibios/eeprom_wb32.c

```diff
@@ -183,36 +205,56 @@ static void WBFLASH_Read(uint32_t ReadAddr, uint8_t *pBuffer, uint16_t NumToRead
  *  the manipulated buffer written after PageErase.
  *******************************************************************************/
 static uint8_t wb_flash_buffer[FEE_PAGE_SIZE] = {0};
-static uint8_t EEPROM_WriteDataByte(uint16_t Address, uint8_t DataByte) {
-    FLASH_Status FlashStatus = FLASH_COMPLETE;
-    uint32_t     raw_address;
-    uint32_t     pagpos;
-    uint16_t     pagoff;
-    uint32_t     offaddr;
+static int EEPROM_WriteDataPage(uint16_t Addr, uint8_t *WriteSrc, size_t Len) {
+    uint16_t     PageReamin;
+    uint16_t     PageOff;
+    uint32_t     PagePos;
+    uint32_t     WriteAddr;
+    uint32_t     WriteLen;
+
     /* exit if desired address is above the limit (e.G. under 2048 Bytes for 4 pages).*/
-    if (Address > FEE_DENSITY_BYTES) {
-        return 0;
+    if (Addr > FEE_DENSITY_BYTES) {
+        eeprom_printf("EEPROM_WriteDataByte(0x%04x) [BAD ADDRESS]\n", Addr);
+        return -1;
     }
 
-    raw_address = Address + FEE_PAGE_BASE_ADDRESS;
-    offaddr     = raw_address - WB32_FLASH_BASE;
     /* calculate which page is affected (Pagenum1/Pagenum2...PagenumN).*/
-    pagpos = offaddr / FEE_PAGE_SIZE;
-    pagoff = offaddr % FEE_PAGE_SIZE;
-    WBFLASH_Read(pagpos * FEE_PAGE_SIZE + WB32_FLASH_BASE, wb_flash_buffer, FEE_PAGE_SIZE);
-    wb_flash_buffer[pagoff] = DataByte;
-    /* Erase the specified FLASH page */
-    FMC_ErasePage(pagpos * FEE_PAGE_SIZE + WB32_FLASH_BASE);
-    /* Clear page latch */
-    FMC_ClearPageLatch();
-
-    for (int i = 0; i < (FEE_PAGE_SIZE >> 2); i++) {
-        FMC->BUF[i] = (*((volatile uint32_t *)(wb_flash_buffer + i * 4)));
+    PagePos    = (Addr + FEE_PAGE_BASE_ADDRESS) / FEE_PAGE_SIZE;
+    PageOff    = (Addr + FEE_PAGE_BASE_ADDRESS) % FEE_PAGE_SIZE;
+    PageReamin = FEE_PAGE_SIZE - PageOff;
+    WriteAddr  = PagePos * FEE_PAGE_SIZE;
+
+    WriteLen = (Len < PageReamin) ? Len : PageReamin;
+
+    if (WriteLen != FEE_PAGE_SIZE) {
```

Comment:
> If a byte is already written,

The function docs **still** dont match the behaviour. All that has been added is an optimisation when writes align to page size, along with a load of duplication.